### PR TITLE
Improve hive text read by allow variable number of fields

### PR DESCRIPTION
**Pull request by KevinyhZou**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
While read data from a hive text file, it would use the first line of hive text file to resize of number of input fields, and sometimes the fields number of first line is not matched with the hive table defined , such as the hive table is defined to have 3 columns, like `test_tbl(a Int32, b Int32, c Int32)`,  but the first line of text file only has 2 fields, and in this suitation, the input fields will be resized to 2, and if the next line of the text file has 3 fields, then the third field can not be read but set a default value 0, which is not right.

And as described above, the hive text file may have variable number of fields,  so we introduce a config `input_format_hive_text_allow_variable_number_of_columns` to allow the file has variable number of columns. 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Review by taiyang-li — APPROVED** [review]
LGTM

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 14882d9a7d7a7da867d1b8626f756e8e0d7e2a21 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62086/14882d9a7d7a7da867d1b8626f756e8e0d7e2a21/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
Missing tests.

**Comment by KevinyhZou** [comment]
The test is added. @alexey-milovidov @Avogar  cc

**Comment by Avogar** [comment]
> While read data from a hive text file, it use a mandatory settings with_names_use_header  with value true, and this config would use the first row of the text file to set its input field names, and sometimes it is mismatched with the hive table defined when the first row's fields number less than the table defines, which lead to some fields can not be read.

From the code it's not correct:
https://github.com/ClickHouse/ClickHouse/blob/68faa1ab9e6e693d3d613ff61fb254c457a8e71e/src/Processors/Formats/Impl/HiveTextRowInputFormat.cpp#L45-L51

As we can see, we don't read names from the data, we just read the first row to determine the number of columns in the data, resize column names accordingly and rollback to the start of first row. Column names in `input_field_names` are set in `StorageHive`. So, we actually never read first row of the data as names header in Hive text. And we can just propagate `with_names=false` to `CSVFormatReader` instead of `true` as it's not because it actually does nothing. And no need to change `with_names_use_header` at all (but need to disable `csv.try_detect_header` to disable header detection).

> use a config csv.allow_variable_number_of_columns for allow the text file has different number fields with hive table defines, and not use the first row of text file to decide the number of fields.

Why not just use setting `input_format_csv_allow_variable_number_of_columns` in the query? Or introduce a similar setting `input_format_hive_text_allow_variable_number_of_columns`? Such changes of default behaviour should be done under a setting.

**Comment by KevinyhZou** [comment]
We encounter such a problem: the hive table is defined to have 3 columns,  like `test_tbl(a Int32, b Int32, c Int32)`, and the hive text file has 2 lines, the  first line has 2 fields, and the second line has 3 fields. If use the logical of `readNames`, then the `input_field_names` would be resized to 2, and this would lead to the second line's third field value can not be read, but set default to 0, which is not right. So the `input_field_names` can not be resized here, so we remove these codes.

And if we just set `with_names=false`,  then the `input_field_names` will not be read into `column_mapping`,  and the `column_mapping` will be constructed on the output header, but sometimes the output reader's schema fields order is not matched the hive table file input fields order,  and this would lead to the file read fields positions reversed. And so we can not just set `with_names=false`.

And we set a config `csv.allow_variable_number_of_columns` to avoid variable number of columns in text file,  if not set this config,  the text read would throw exceptions if the column number is variable. and set it here, I think can make it easy to use, not need to add a extra config while query.

**Comment by Avogar** [comment]
Ok, so, we only want to allow variable number of columns for HiveText input format (please, edit the PR description). Still I think it will be good to add a setting `input_format_hive_text_allow_variable_number_of_columns` and enable it by default, so the old behaviour can be restored by disabling this setting.

**Comment by KevinyhZou** [comment]
OK

**Comment by KevinyhZou** [comment]
cc @Avogar

File: src/Core/SettingsChangesHistory.h [hunk 1]
```diff
@@ -125,6 +125,7 @@ static std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> sett
               {"azure_max_upload_part_size", 5ull*1024*1024*1024, 5ull*1024*1024*1024, "The maximum size of part to upload during multipart upload to Azure blob storage."},
               {"azure_upload_part_size_multiply_factor", 2, 2, "Multiply azure_min_upload_part_size by this factor each time azure_multiply_parts_count_threshold parts were uploaded from a single write to Azure blob storage."},
               {"azure_upload_part_size_multiply_parts_count_threshold", 500, 500, "Each time this number of parts was uploaded to Azure blob storage, azure_min_upload_part_size is multiplied by azure_upload_part_size_multiply_factor."},
+              {"input_format_hive_text_allow_variable_number_of_columns", true, true, "Ignore extra columns in Hive Text input (if file has more columns than expected) and treat missing fields in Hive Text input as default values."},
```
Comment (Avogar) [inline src/Core/SettingsChangesHistory.h:128]:
```suggestion
              {"input_format_hive_text_allow_variable_number_of_columns", false, true, "Ignore extra columns in Hive Text input (if file has more columns than expected) and treat missing fields in Hive Text input as default values."},
```
Change is `false` -> `true` as previously it didn't work

Comment (KevinyhZou) [reply src/Core/SettingsChangesHistory.h]: done

**Review by Avogar — APPROVED** [review]

File: src/Core/SettingsChangesHistory.h (hunk 1 above)
Comment (Algunenano) [reply src/Core/SettingsChangesHistory.h]: This setting has been added to the wrong version: https://s3.amazonaws.com/clickhouse-test-reports/62086/44b3ce9ec88ac147e33856170c402bac3444bb54/upgrade_check__debug_.html

Comment (KevinyhZou) [reply src/Core/SettingsChangesHistory.h]: OK， thanks for your fix

**Comment by tavplubix** [comment]
@Avogar, please check the CI status

**Comment by clickgapai** [comment]
Hi @KevinyhZou @Avogar @taiyang-li — while reviewing this PR I found the following:

- **Coverage gap:** `input_format_hive_text_allow_variable_number_of_columns=0` (strict mode) and the extra-columns branch are untested for HIVETEXT — https://github.com/ClickHouse/ClickHouse/pull/103981

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**